src/liteclient.cc

```diff
@@ -205,6 +207,117 @@ boost::container::flat_map<int, Uptane::Target> LiteClient::getTargets() {
   return sorted_targets;
 }
 
+// Update a device to the given Target
+data::ResultCode::Numeric LiteClient::update(Uptane::Target& target, bool force_update) {
+  data::ResultCode::Numeric update_result = data::ResultCode::Numeric::kUnknown;
+  std::string download_reason;
+  auto current_target = getCurrent();
+
+  Uptane::Target target_to_update{target};
+
+  UpdateType update_type = UpdateType::kNewTargetUpdate;
+  if (force_update) {
+    update_type = UpdateType::kTargetForcedUpdate;
+  } else if (target.filename() == current_target.filename() || !isTargetValid(target)) {
+    update_type = UpdateType::kCurrentTargetSync;
+    target_to_update = current_target;
+  }
+
+  // set an update correlation ID to "bind" all report events generated during a single update together
+  // so the backend can know to which of the updates specific events are related to
+  Target::setCorrelationID(target_to_update);
+
+  // prepare for an update
+  switch (update_type) {
+    case UpdateType::kNewTargetUpdate: {
+      logTarget("Updating Active Target: ", current_target);
+      logTarget("To New Target: ", target_to_update);
+      download_reason = "Updating from " + current_target.filename() + " to " + target_to_update.filename();
+
+      appsInSync();
+      break;
+    }
+    case UpdateType::kCurrentTargetSync: {
+      logTarget("Syncing current Target: ", target_to_update);
+      download_reason = "Syncing current Target " + target_to_update.filename();
+
+      if (appsInSync() && target_to_update.sha256Hash() == current_target.sha256Hash()) {
+        setAppsNotChecked();
+        return data::ResultCode::Numeric::kAlreadyProcessed;
+      }
+
+      break;
+    }
+    case UpdateType::kTargetForcedUpdate: {
+      logTarget("Updating to Target: ", target);
+      download_reason = "Manual update to Target " + target.filename();
+      break;
+    }
+    default: {
+      // this shouldn't really happen at the runtime
+      throw std::runtime_error("Unsupported type of update:  " + std::to_string(update_type));
+    }
+  }
+
+  // do update, i.e. download and install
+  do {
+    update_result = download(target_to_update, download_reason);
+
+    // TODO: it should be really just `download` and `install`
+    if (update_result != data::ResultCode::Numeric::kOk) {
+      break;
+    }
+
+    if (VerifyTarget(target_to_update) != TargetStatus::kGood) {
+      // why do we do it here not inside of packman->download() ???
+      data::InstallationResult res{data::ResultCode::Numeric::kVerificationFailed, "Downloaded target is invalid"};
+      // why do we send notifyInstallFinished if notifyInstallStarted hasn't been called ?
+      notifyInstallFinished(target_to_update, res);
```

Comment:
The given code sends `EcuInstallationCompleted` event https://github.com/foundriesio/device-gateway/blob/1aa14557902b372e7e787c1d4ffd8358dda48006/storage/db.go#L285 what makes backend to store status "Failed installing {Target}" while installation actually even has not started and EcuInstallationStarted has been send (btw DG doesn't handle the given event at all?). 
So, I think the correct behavior is to send just EcuDownloadCompleted event which is sent anyway in LiteClient::download(), so the backedn/DG/fioctl will report "Failed downloading " instead of "Failed installing"

That verification in libaktualizr makes sense only for the file package manager, it's used for verification of the downloaded file integrity and authenticity. Not sure if it makes sense to do for the ostree, as libostree verifies objects checksums/hashes internally during commit pulling process...

Reply:
+1